[mingshan2705] iP

mingshan2705 opened the pull request:
# Implementation of Yapper Chatbot 📝
> 🗣️ **Yapper is here to help!**  
> Start yapping with commands like `todo`, `deadline`, and `event`!

This PR introduces the **Yapper** chatbot, an interactive task management system. Below are the key enhancements:

## Features Implemented
- [x] `todo` command for adding tasks 📌
- [x] `deadline` command with date/time support ⏳
- [x] `event` command with start and end times 📅
- [x] `list` command to display tasks 📋
- [x] `find` command to search tasks 🔍

## How to Use
1. download the `.jar` file [here](https://github.com/mingshan2705/ip/releases)
2. run the JAR file using `java -jar yapper.jar`

## Storage
Your task list is stored as a text file under `./data/yapper.txt`. You can change this in the `main` method under `Yapper.java`.

```java
public static void main(String[] args) {
    new Yapper("./data/yapper.txt").run();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

mingshan2705 committed:
Change Duke to Yapper

mingshan2705 committed:
Add echo feature

mingshan2705 committed:
Add support for list

mingshan2705 committed:
Add Task class

mingshan2705 committed:
Add Mark as Done functionality

mingshan2705 committed:
Add Todo class

mingshan2705 committed:
Add Deadline class

mingshan2705 committed:
Add Event class

mingshan2705 committed:
Add support for tracking different tasks

mingshan2705 committed:
Modify runtest scripts

mingshan2705 committed:
Modify input and expected output

mingshan2705 committed:
Update tests

mingshan2705 committed:
Add support for dealing with errors

mingshan2705 committed:
Update test input and expected output

mingshan2705 committed:
Add support for deleting tasks

mingshan2705 committed:
Add Storage class to handle file load and save functionality

mingshan2705 committed:
Integrate Storage with Yapper

mingshan2705 committed:
Upload data files for observability

mingshan2705 committed:
Merge branch-level-7 into master

mingshan2705 committed:
Add datetime feature for Deadline and Event

mingshan2705 committed:
Update tests to accomdate datetime changes

mingshan2705 committed:
Merge branch-level-8 into master

mingshan2705 committed:
Add Commands

mingshan2705 committed:
Add Parser, Ui, and TaskList

mingshan2705 committed:
Update Storage and Yapper to integrate with changes

mingshan2705 committed:
Update tests with new features

mingshan2705 committed:
Merge branch-A-MoreOOP into master

mingshan2705 committed:
Reorganise packages

mingshan2705 committed:
Merge add-gradle-support into master

mingshan2705 committed:
Add JUnit tests

mingshan2705 committed:
Modify build.gradle for JAR File creation

mingshan2705 committed:
Add search functionality

mingshan2705 committed:
Add JavaDoc comments

mingshan2705 committed:
Ensure coding standard

mingshan2705 committed:
Merge branch 'branch-A-JavaDoc'

mingshan2705 committed:
Merge branch 'branch-A-CodingStandard'

mingshan2705 committed:
Fix unsupported toList bug

mingshan2705 committed:
Add UI for Yapper

mingshan2705 committed:
Merge branch Level-10 into master

quangthangisme commented on src/main/java/yapper/Parser.java:
```diff
@@ -0,0 +1,45 @@
+package yapper;
+
+import yapper.command.*;
```
You should avoid wildcard imports.

quangthangisme commented on src/main/java/yapper/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+import java.io.IOException;
+
+public abstract class Command {
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws IOException;
```
If I understand correctly, the command execution currently includes both printing via Ui and saving via Storage. Would it be better to create a Data Transfer Object to pass the feedback message to Ui and the tasks to Storage separately?

quangthangisme commented on src/main/java/yapper/Storage.java:
```diff
@@ -0,0 +1,93 @@
+            Event event = (Event) task;
+            return String.join(" | ", taskType, status, event.description, event.from.toString(), event.to.toString());
+        }
+        throw new IllegalArgumentException("Unknown task type: " + task.getClass().getSimpleName());
```
It seems like you're relying too much on `IllegalArgumentException`. It might be better to create more specific exceptions, such as `UnknownCommandException` and `InvalidArgumentException`.

quangthangisme commented on src/main/java/yapper/Storage.java:
```diff
@@ -0,0 +1,93 @@
+            return String.join(" | ", taskType, status, deadline.description, deadline.by.toString());
+        } else if (task instanceof Event) {
+            Event event = (Event) task;
+            return String.join(" | ", taskType, status, event.description, event.from.toString(), event.to.toString());
```
This line exceeds the character limit of 100. You could split it into multiple lines.

quangthangisme commented on src/main/java/yapper/task/Event.java:
```diff
@@ -0,0 +1,39 @@
+ * Extends the {@link Task} class to include additional details for event timing.
+ */
+public class Event extends Task {
+    public LocalDateTime from;
```
You could use a more noun-like attribute name, such as `startTime`, for better clarity.

quangthangisme reviewed (commented):
Overall, I think your code is clean and well-structured. However, there are some coding stardard violations you might want to address. I have also provided comments on areas where I believe your code can be improved.

danplatypus30 commented on src/main/java/yapper/command/FindCommand.java:
```diff
@@ -0,0 +1,37 @@
+        } else {
+            StringBuilder message = new StringBuilder("Here are the matching tasks in your list:");
+            for (int i = 0; i < matchingTasks.size(); i++) {
+                message.append("\n ").append(i + 1).append(".").append(matchingTasks.get(i));
```
Perhaps you can separate long lines like this into multiple lines?

danplatypus30 commented on src/main/java/yapper/command/ListCommand.java:
```diff
@@ -0,0 +1,32 @@
+package yapper.command;
+
+import yapper.*;
```
Perhaps instead of using wildcard *, you can list specific classes out just for this mod?

danplatypus30 reviewed (commented):
Overall I like that your code is really well formatted and packaged nicely :)

danplatypus30 commented on src/main/java/yapper/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+ * Represents a command to add a deadline task to the task list.
+ * Parses the user input to create a {@link Deadline} and adds it to the {@link TaskList}.
+ */
+public class AddDeadlineCommand extends Command {
```
I like that you abstracted these commands into multiple classes, however I feel like these commands do not really need to be in separate classes

danplatypus30 commented on src/main/java/yapper/Yapper.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void start(Stage stage) throws IOException {
+        storage = new Storage(FILE_PATH);
```
Perhaps you could use this.storage instead?

danplatypus30 reviewed (commented):
A little more comments to hit 4 comments

mingshan2705 committed:
Ensure code quality

Ensured code quality by following SLAP

mingshan2705 committed:
Add assertions

Added assertions such that the start time and end time is not null

mingshan2705 committed:
Merge pull request #1 from mingshan2705/branch-A-Assertions

Add assertions

mingshan2705 committed:
Merge

mingshan2705 committed:
Merge pull request #2 from mingshan2705/branch-A-CodeQuality

Ensure code quality

mingshan2705 committed:
Add View Schedule Functionality

Add a command `schedule YYYY-MM-DD' for users to view the tasks for a
specific date.

This allows the user to filter the tasklist and focus on the tasks for a
certain day.

mingshan2705 committed:
Merge branch 'branch-B-ViewSchedules'

mingshan2705 committed:
Merge branch 'master' of https://github.com/mingshan2705/ip
Merge branch ViewSchedule into master

mingshan2705 committed:
Add user guide

mingshan2705 committed:
Fix missing JavaFX runtime components error
